aws/data_source_aws_s3_bucket_objects_test.go

```diff
@@ -0,0 +1,327 @@
+					testAccCheckAwsS3ObjectsDataSourceExists("data.aws_s3_bucket_objects.yesh"),
+					resource.TestCheckResourceAttr("data.aws_s3_bucket_objects.yesh", "keys.#", "1"),
+					resource.TestCheckResourceAttr("data.aws_s3_bucket_objects.yesh", "keys.0", "arch/three_gossips/turret"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccDataSourceAWSS3BucketObjects_fetchOwner(t *testing.T) {
+	rInt := acctest.RandInt()
+	basic := testAccAWSDataSourceS3ObjectsConfigOwners(rInt)
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:                  func() { testAccPreCheck(t) },
+		Providers:                 testAccProviders,
+		PreventPostDestroyRefresh: true,
+		Steps: []resource.TestStep{
+			{
+				Config: basic,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsS3ObjectsDataSourceExists("data.aws_s3_bucket_objects.yesh"),
+					resource.TestCheckResourceAttr("data.aws_s3_bucket_objects.yesh", "keys.#", "2"),
+					resource.TestCheckResourceAttr("data.aws_s3_bucket_objects.yesh", "owners.#", "2"),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckAwsS3ObjectsDataSourceExists(addr string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[addr]
+		if !ok {
+			return fmt.Errorf("Can't find S3 objects data source: %s", addr)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("S3 objects data source ID not set")
+		}
+
+		return nil
+	}
+}
+
+func testAccAWSDataSourceS3ObjectsConfigResources(randInt int) string {
+	return fmt.Sprintf(`
+resource "aws_s3_bucket" "objects_bucket" {
+  bucket = "tf-objects-test-bucket-%d"
+}
+
+resource "aws_s3_bucket_object" "object1" {
+  bucket  = "${aws_s3_bucket.objects_bucket.id}"
+  key     = "arch/three_gossips/turret"
+  content = "Delicate"
+}
+
+resource "aws_s3_bucket_object" "object2" {
+  bucket  = "${aws_s3_bucket.objects_bucket.id}"
+  key     = "arch/three_gossips/broken"
+  content = "Dark Angel"
+}
+
+resource "aws_s3_bucket_object" "object3" {
+  bucket  = "${aws_s3_bucket.objects_bucket.id}"
+  key     = "arch/navajo/north_window"
+  content = "Balanced Rock"
+}
+
+resource "aws_s3_bucket_object" "object4" {
+  bucket  = "${aws_s3_bucket.objects_bucket.id}"
+  key     = "arch/navajo/sand_dune"
+  content = "Queen Victoria Rock"
+}
+
+resource "aws_s3_bucket_object" "object5" {
+  bucket  = "${aws_s3_bucket.objects_bucket.id}"
+  key     = "arch/partition/park_avenue"
+  content = "Double-O"
+}
+
+resource "aws_s3_bucket_object" "object6" {
+  bucket  = "${aws_s3_bucket.objects_bucket.id}"
+  key     = "arch/courthouse_towers/landscape"
+  content = "Fiery Furnace"
+}
+
+resource "aws_s3_bucket_object" "object7" {
+  bucket  = "${aws_s3_bucket.objects_bucket.id}"
+  key     = "arch/rubicon"
+  content = "Devils Garden"
+}
+`, randInt)
+}
+
+func testAccAWSDataSourceS3ObjectsConfigBasic(randInt int) string {
+	return fmt.Sprintf(`
+%s
+
+data "aws_s3_bucket_objects" "yesh" {
+  bucket    = "${aws_s3_bucket.objects_bucket.id}"
```

Comment:
I'm guessing you were lucking out with how this acceptance testing was working since currently the data source has no dependencies on the resources and it would be a race condition to when the data source would get called. 😄 It seems like we'd need to setup all the resources and data source to referencing the "previous" resource to ensure all the resources were created before the data source was called, e.g.

```hcl
# ...
resource "aws_s3_bucket_object" "object7" {
  bucket  = "${aws_s3_bucket_object.object6.bucket}" # or use depends_on = [aws_s3_bucket_object.object1, aws_s3_bucket_object.object2, ...]
  key     = "arch/rubicon"
  content = "Devils Garden"
}

data "aws_s3_bucket_objects" "yesh" {
  bucket    = "${aws_s3_bucket_object.object7.bucket}"
# ...
```

Reply:
Since the tests use the same setup, the `%s` gets replaced with the resources. 😄 
```go
return fmt.Sprintf(`
%s #<===

data "aws_s3_bucket_objects" "yesh" {
  bucket    = "${aws_s3_bucket.objects_bucket.id}"
  prefix    = "arch/navajo/"
  delimiter = "/"
}
`, testAccAWSDataSourceS3ObjectsConfigResources(randInt)) //<===
```

Reply:
Right, but how are the dependencies ordered in the expanded Terraform configuration? Terraform does not perform API operations top down based on the configuration. 😄 The data source has no ordering directives (explicitly with `depend_on` (which doesn't work well with data sources anyways) or implicitly with attribute references) to tell Terraform that the resources must be created first.